flox/core.py

```diff
@@ -1784,4 +1784,9 @@ def groupby_reduce(
 
     if _is_minmax_reduction(func) and is_bool_array:
         result = result.astype(bool)
+
+    # we lose dtype information when creating a pandas Index
```

Comment:
Nice, this was a real bug it seems like.

Reply:
Actually this is a bad idea. The core issue is that for `cohorts` and `blockwise` we use `np.unique` and concatenate, so dtype is preserved. For `map-reduce` we convert to a pandas Index so it always casts. (this is when expected_groups is None).

I think the solution is to always cast to whatever type pandas likes until we can specify dtypes for pd.Index